File: 2022_07_22_code_snippets/TypesScript/misc/Queue.ts
type List<T> = [head: T, tail: List<T>] | null;

var listEmpty: List<number> = null;
var list123: List<number> = [1, [2, [3, null]]];

function rev<T>(l: List<T>, acc: List<T> = null): List<T> {
  return l === null ? acc : rev(l[1], [l[0], acc]);
}

console.log(rev(listEmpty)); // null
console.log(rev(list123)); // [3, [2, [1, null]]]

type Queue<T> = {
  front: List<T>;
  back: List<T>;
}

var queueEmpty: Queue<number> = { front: listEmpty, back: null };
var queue123: Queue<number> = { front: list123, back: null };

function peek<T>(q: Queue<T>) {
  return q.front === null ? null : q.front[0];
}

console.log(peek(queueEmpty)); // null
console.log(peek(queue123)); // 1

function enqueue<T>(q: Queue<T>, x: T): Queue<T> {
  return q.front === null ? { front: [x, null], back: null } : { ...q, back: [x, q.back] };
}

console.log(enqueue(queueEmpty, 1)); // { front: [ 1, null ], back: null } 
console.log(enqueue(enqueue(queueEmpty, 1), 2)); // { front: [ 1, null ], back: [ 2, null ] }

function denqueue<T>(q: Queue<T>): Queue<T> | null {
  return q.front === null ? null
    : q.front[1] === null ? { front: rev(q.back), back: null }
      : { ...q, front: q.front[1] }
}

var queue1234 = [1, 2, 3, 4].reduce(enqueue, queueEmpty);

console.log(queue1234); // {front:[1,null],back:[4,[3,[2,null]]]}
console.log(denqueue(queue1234)); // "{front:[2,[3,[4,null]]],back:null}
console.log(denqueue(denqueue(queue1234)!)); // "{front":[3,[4,null]],back:null}

[denqueue, denqueue, denqueue, denqueue, denqueue].reduce((q, f) => f(q)!, queue1234); // { "front": null, "back": null }
[denqueue, denqueue, denqueue, denqueue, denqueue, denqueue].reduce((q, f) => f(q)!, queue1234); null

File: 2023_09_02_algorithms_and_data_structures_in_fsharp/AlgorithmsAndDataStructures/Queue.fs
module AlgorithmsAndDataStructures.Queue

open Utils

// https://www.youtube.com/watch?v=rCiBfZO67A4&t=228s Functional Queues | OCaml Programming | Chapter 5 Video 7


type Queue<'a> = { Front: 'a list; Back: 'a list }

let empty = { Front = []; Back = [] }

let peek =
    function
    | { Front = [] } -> None
    | { Front = x :: _ } -> Some x

peek empty === None


let enqueue x =
    function
    | { Front = [] } -> { Front = [ x ]; Back = [] }
    | q -> { q with Back = x :: q.Back }

let denqueue =
    function
    | { Front = [] } -> None
    | { Front = _ :: []; Back = back } -> Some { Front = List.rev back; Back = [] }
    | { Front = _ :: t; Back = back } -> Some { Front = t; Back = back }


denqueue empty === None

let q = empty |> enqueue 5 |> enqueue 10 |> enqueue 15

peek q === Some 5
q |> denqueue |> Option.bind peek === Some 10
q |> denqueue |> Option.bind denqueue |> Option.bind denqueue === Some empty







// https://www.lucc.pl/inf/struktury_danych_i_zlozonosc_obliczeniowa/cormen_-_wprowadzenie_do_algorytmow.pdf

// https://dahlan.unimal.ac.id/files/ebooks/2013%20Algorithms_Unlocked.pdf
// https://doc.lagout.org/programmation/Functional%20Programming/Chris_Okasaki-Purely_Functional_Data_Structures-Cambridge_University_Press%281998%29.pdf
// https://doc.lagout.org/programmation/

// https://cs3110.github.io/textbook/cover.html
